bgbb/bgbb_utils.py

```diff
@@ -0,0 +1,120 @@
+from collections import OrderedDict
+from itertools import count
+from typing import List, Optional, Sequence, Tuple
+
+import numpy as np
+
+# sim_dat = gen_buy_die(n_opps, n_users, abgd=SimPars.lst, p_th=[pp, th])
+import numpy.random as nr
+
+from numba import njit
+from pandas import DataFrame
+
+Prob = float
+
+
+###############
+# Simulations #
+###############
+def _gen_probs(abgd: List[float], n=10):
+    pa, pb, ta, tb = abgd
+    p = nr.beta(pa, pb, size=n)
+    th = nr.beta(ta, tb, size=n)
+    return p, th
+
+
+def gen_buy_die(
+    n_opps,
+    n_users,
+    abgd: List[float],
+    p_th: Optional[Tuple[Sequence[Prob], Sequence[Prob]]] = None,
+    n_opps_name="n_opps",
+    seed=0,
+):
+    """
+    """
+    nr.seed(seed)
+    if p_th is None:
+        raise NotImplementedError
+        # to dict
+        p, th = _gen_probs(abgd, n=n_users)
+    else:
+        p, th = p_th
+    txs, xs = np.empty_like(p, dtype=np.int), np.empty_like(p, dtype=np.int)
+    bern = lambda p: nr.binomial(1, p, size=n_opps)
+
+    for i, pi, thi in zip(count(), p, th):
+        buys, dies = bern(pi), bern(thi)
+        xs[i], txs[i] = get_x_tx(buys, dies)
+
+    ret = DataFrame(
+        OrderedDict([("p", p), ("th", th), ("frequency", xs), ("recency", txs)])
+    ).assign(**{n_opps_name: n_opps})
+    return ret
+
+
+@njit
+def get_x_tx(buys, dies) -> Tuple[int, int]:
+    """
+    Converts simulated 'buy'/'die' events into frequency/recency
+    counts.
+    """
+    x = 0
+    tx = 0
+    for opp, buy, die in zip(range(1, len(buys) + 1), buys, dies):
+        if die:
+            break
+        if buy:
+            x += 1
+            tx = opp
+    return x, tx
+
+
+class AbgdParams:
```

Comment:
I think you might be looking for the namedtuple collection here. It lets you define a tuple with names.

```python
from collections import namedtuple

AbgdParams = namedtuple("AbgdParams", ["a", "b", "g", "d"])

# constructing a new tuple
x = AbgdParams(1, 2, 3, 4)

# desugar
a, b, g, d = x

print(a, b, g, d)
# (1, 2, 3, 4)

# return an ordered dictionary
x._asdict()
# OrderedDict([('a', 1), ('b', 2), ('g', 3), ('d', 4)])